include/watchdog.h

```diff
@@ -1,116 +1,206 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
+ * Copyright (c) 2017 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
+/**
+ * @brief Enable calling application callback function on watchdog timeout.
+ */
+#define WDT_MODE_CALLBACK	BIT(0)
 
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
+/**
+ * @brief Pause watchdog timer when CPU is in sleep state.
+ */
+#define WDT_PAUSE_IN_SLEEP	BIT(1)
+
+/**
+ * @brief Pause watchdog timer when CPU is halted by the debugger.
+ */
+#define WDT_PAUSE_HALTED_BY_DBG		BIT(2)
+
+/**
+ * @brief Watchdog timeout window.
+ *
+ * Each installed timeout needs reloading in specified time window, otherwise
+ * the watchdog will trigger.
+ *
+ * @param up Upper limit of watchdog reload timeout in miliseconds.
+ * @param down Lower limit of watchdog reload timeout in miliseconds.
+ */
+struct wdt_window {
+	u16_t up;
+	u16_t down;
```

Comment:
@mnkp Could you explain why signed is used in kernel API. Using signed here, where negative values makes no sense is at least weird for me.

Reply:
Good question. I'm afraid I cannot give any sensible answer here. Maybe someone else?

Reply:
isn't it so because there is no garanty the timer will exactly be stopped at the right tick, so the remaining time might be below 0 (which then is a stop condition for the timer as well as if it would be 0)